### ISSUE TITLE
Не применяется BEMHTML-шаблон

### ISSUE THREAD

**Issue by varya** [bem/bem-forum-content-ru#240]
Спасибо форуму, сборка происходит нормально. Но результатов BEMHTML на клиенте пока нет.

В JavaScript код такой:
```
        var reposBEMJSON = this.repos.map(function(repo) {

            return {
                block: 'test',
                //elem: 'repo',
                content: '1111',
                info: {
                    name: repo.name,
                    url: repo.html_url
                }
            }
        });

        var html = BEMHTML.apply(reposBEMJSON)
        console.log(html);
```
https://github.com/varya/varya.github.com/blob/features/github-repos/desktop.blocks/github/github.js#L49

Зависимости по технологиям указываю:  https://github.com/varya/varya.github.com/blob/features/github-repos/desktop.blocks/github/github.deps.js#L12

Судя по коду index.min.js шаблон всё-таки включен в сборку:

```
if(!__$ctx.elem&&__$ctx.block==="test"&&__$ctx._mode==="content"){__$ctx.__$a=0;return"XXXX"}
```
Но в консоль выводится `undefined`. В браузер приезжает свежий, незакешированной `index.min.js` — я проверила.

В чём может быть проблема и как дебажить дальше?

**Comment by tadatuta** [comment]
@varya не хватает зависимости от блока `i-bem` в технологии `bemhtml`

**Comment by varya** [comment]
Хм, не очень понятно.

Я добавила зависимость к блоку `test`:

```
[{

    tech: 'bemhtml',
    mustDeps: [
      { block: 'i-bem' }
    ]

}]
```

Но это не помогло. Помогла только зависимость для блока `github` (для которого и пишу JavaScript):

```
{

    tech: 'js',
    mustDeps: [
      { tech: 'bemhtml', block: 'test' }
    ]

}, {
    tech: 'js',
    mustDeps: { tech: 'bemhtml', block: 'i-bem' }

}]
```

Как-то нелогично. Почему первый вариант не подходит?

**Comment by varya** [comment]
А можно ещё раз, что от чего должно зависеть? Я указала i-bem https://github.com/varya/varya.github.com/blob/features/github-repos/desktop.blocks/github/github.deps.js#L12 Теперь я получаю не `undefined`, но всё равно результат дефолтных щаблонов:

```
<div class="test">1111</div><div class="test">1111</div>
```

А хочу, чтобы применился шаблон на блок test: https://github.com/varya/varya.github.com/blob/features/github-repos/desktop.blocks/test/test.bemhtml

**Comment by tadatuta** [comment]
@varya прислал PR: https://github.com/varya/varya.github.com/pull/63/files с пояснениями по диффу

**Comment by varya** [comment]
Спасибо, всё заработало! А дальше уже переделать на зависимость от собственного элемента оказалось несложно https://github.com/varya/varya.github.com/commit/410b359d0dc7f9abc8bbe321439a304651a9d2bd

**Comment by varya** [comment]
Но дальше всё равно не понятно :-)

Теперь я хочу использовать в своём шаблоне блок link. Почему вот такой зависимости недостаточно ("ссылка" строится по дефолтным шаблонам): https://github.com/varya/varya.github.com/commit/bc4f0ef1fb2eae62b57f1c2d6f81fd713bb42dd6

**Comment by tadatuta** [comment]
потому что никто не зовет шаблон ссылки в клиентские шаблоны. должно быть:
```js
[{
    mustDeps: 'i-bem'
 },
{
    tech: 'js',
    shouldDeps: { block: 'link', tech: 'bemhtml' }
}]
```

**Comment by varya** [comment]
А почему?? Ведь блок `github` независим, он должен строиться по-полному, и без разницы, откуда его позвали: из JavaSscript или BEMHTML. /cc @veged @dfilatov

**Comment by tadatuta** [comment]
Идея `depsByTech` предполагает, что ты можешь явно управлять тем, какие из шаблонов попадут в клиентский JS (иначе пришлось бы собирать все).

Когда блоки задекларированы без указания технологии, они собираются «один в один». В данном случае шаблон `link` попадает в `index.bemhtml.js`, но не попадает в `index.min.js` пока это не будет указано явно.

При этом нет необходимости из блока `github` указывать зависимость depsByTech от `i-bem`, сборщик поймет, что хотя `link` зависит от нее «обычным способом», она все равно нужна на клиенте.

**Comment by varya** [comment]
То есть если я использую библиотеку, в которой блок A строится с использованием блока B, и вдруг решила построить это всё на клиенте, мне нужно будет не только прописать зависимость depsByTech от блока A, но так же от блока B (и от всех других блоков, которые требуются для работы A)?

**Comment by tadatuta** [comment]
нет, если в блоке А прописана зависимость от В, то на клиент достаточно будет прописать только А.

**Comment by varya** [comment]
Но эта "зависимость A от B" должна быть byTech или обычная тоже сойдет?

**Comment by apsavin** [comment]
Должно работать с обычной.

> При этом нет необходимости из блока github указывать зависимость depsByTech от i-bem, сборщик поймет, что хотя link зависит от нее «обычным способом», она все равно нужна на клиенте.

**Comment by varya** [comment]
Тогда возвращаюсь к своему вопросу "Почему вот такой зависимости не достаточно? https://github.com/varya/varya.github.com/commit/bc4f0ef1fb2eae62b57f1c2d6f81fd713bb42dd6"

**Comment by apsavin** [comment]
У тебя было указано в зависимостях через depsByTech, что шаблон github__repo.bemhtml должен попасть на клиент?

**Comment by varya** [comment]
Да, вот https://github.com/varya/varya.github.com/blob/410b359d0dc7f9abc8bbe321439a304651a9d2bd/desktop.blocks/github/github.deps.js#L9

**Comment by zxqfox** [comment]
See also: #111

**Comment by apsavin** [comment]
@zxqfox У Вари и так массив, разве нет?

**Comment by tadatuta** [comment]
@varya упячка [вот здесь](https://github.com/varya/varya.github.com/blob/410b359d0dc7f9abc8bbe321439a304651a9d2bd/desktop.blocks/i-bem/i-bem.deps.js) — из-за того, что тут используется `mustDeps`, многие блоки (и `link` в том числе) оказываются раньше, чем базовые шаблоны. соответственно, базовые шаблоны «перевешивают» шаблоны блоков.

**Comment by varya** [comment]
@tadatuta, Да вроде нет. Если переделать на shouldDeps — ничего не меняется, кроме того что highlight перестаёт работать как задумано :-)

**Comment by zxqfox** [comment]
@apsavin Просто ссылку оставил ;) вообще, я другой пост искал, и не факт, что он на форуме. Там где @veged, @floatdrop и @matmuchrapna обсуждали депсы.

**Comment by tadatuta** [comment]
@varya проверил, у меня после
```diff
diff --git a/desktop.blocks/i-bem/i-bem.deps.js b/desktop.blocks/i-bem/i-bem.deps.js
index 18e5f14..a97221a 100644
--- a/desktop.blocks/i-bem/i-bem.deps.js
+++ b/desktop.blocks/i-bem/i-bem.deps.js
@@ -1,5 +1,5 @@
 ({
-    mustDeps: [
+    shouldDeps: [
         {
             block: 'highlight',
             elem: 'hack'
```

шаблоны заработали как задумано.

**Comment by varya** [comment]
Спасибо.
Я решила по-другому (но суть та же):
Поскольку `highlight__hack` мне нужен рано, оставила `mustDeps`, но сделала это зависимостью `page`, а не `i-bem`.
А у блока `github` прошу `i-bem` как `mustDeps`, чтобы он гарантированно был раньше.
